File: python/FileActions.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the feather-tk project.

import opentimelineio as otio
import ftkPy as ftk
import ibisPy as ibis

import weakref

class Actions:
    """
    File actions.
    """
    def __init__(self, context, app, mainWindow):

        appWeak = weakref.ref(app)
        mainWindowWeak = weakref.ref(mainWindow)
        self.actions = {}

        self.actions["New"] = ftk.Action(
            "New",
            "FileNew",
            ftk.KeyShortcut(ftk.Key.N, ftk.commandKeyModifier),
            lambda: appWeak().documentModel().newDocument())
        self.actions["New"].tooltip = "Create a new file."

        self.actions["Open"] = ftk.Action(
            "Open",
            "FileOpen",
            ftk.KeyShortcut(ftk.Key.O, ftk.commandKeyModifier),
            lambda: context.getSystemByName("ftk::FileBrowserSystem").open(
                mainWindowWeak(),
                appWeak().documentModel().open))
        self.actions["Open"].tooltip = "Open a file."

        self.actions["Close"] = ftk.Action(
            "Close",
            "FileClose",
            ftk.KeyShortcut(ftk.Key.E, ftk.commandKeyModifier),
            lambda: appWeak().documentModel().close(mainWindowWeak().currentTab()))
        self.actions["Close"].tooltip = "Close the current file."

        self.actions["CloseAll"] = ftk.Action(
            "Close All",
            "FileCloseAll",
            ftk.KeyShortcut(ftk.Key.E, ftk.commandKeyModifier, ftk.KeyModifier.Shift),
            lambda: appWeak().documentModel().closeAll())
        self.actions["CloseAll"].tooltip = "Close all files."

        self.actions["Exit"] = ftk.Action(
            "Exit",
            ftk.KeyShortcut(ftk.Key.Q, ftk.commandKeyModifier),
            lambda: appWeak().exit())

        selfWeak = weakref.ref(self)
        self._documentsObserver = ibis.models.DocumentListObserver(
            app.documentModel().observe,
            lambda documents: selfWeak()._documentsCallback(documents))

    def _documentsCallback(self, documents):
        self.actions["Close"].enabled = len(documents) > 0
        self.actions["CloseAll"].enabled = len(documents) > 0

File: bin/FileActions.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the ibis project.

#include "FileActions.h"

#include "App.h"
#include "MainWindow.h"

namespace ibis
{
    void FileActions::_init(
        const std::shared_ptr<ftk::Context>& context,
        const std::shared_ptr<App>& app,
        const std::shared_ptr<MainWindow>& mainWindow)
    {
        auto appWeak = std::weak_ptr<App>(app);
        _actions["New"] = ftk::Action::create(
            "New",
            "FileNew",
            ftk::KeyShortcut(ftk::Key::N, ftk::commandKeyModifier),
            [appWeak]
            {
                appWeak.lock()->newDocument();
            });
        _actions["New"]->setTooltip("Create a new file.");

        _actions["Open"] = ftk::Action::create(
            "Open",
            "FileOpen",
            ftk::KeyShortcut(ftk::Key::O, ftk::commandKeyModifier),
            [appWeak]
            {
                appWeak.lock()->open();
            });
        _actions["Open"]->setTooltip("Open a file.");

        _actions["Save"] = ftk::Action::create(
            "Save",
            "FileSave",
            ftk::KeyShortcut(ftk::Key::S, ftk::commandKeyModifier),
            [appWeak]
            {
                appWeak.lock()->save();
            });
        _actions["Save"]->setTooltip("Save the current file.");

        std::weak_ptr<MainWindow> mainWindowWeak(mainWindow);
        _actions["Close"] = ftk::Action::create(
            "Close",
            "FileClose",
            ftk::KeyShortcut(ftk::Key::E, ftk::commandKeyModifier),
            [appWeak, mainWindowWeak]
            {
                appWeak.lock()->getDocumentModel()->close(
                    mainWindowWeak.lock()->getCurrentTab());
            });
        _actions["Close"]->setTooltip("Close the current file.");

        _actions["CloseAll"] = ftk::Action::create(
            "Close All",
            "FileCloseAll",
            ftk::KeyShortcut(ftk::Key::E, ftk::commandKeyModifier, ftk::KeyModifier::Shift),
            [appWeak]
            {
                appWeak.lock()->getDocumentModel()->closeAll();
            });
        _actions["CloseAll"]->setTooltip("Close all files.");

        _actions["Exit"] = ftk::Action::create(
            "Exit",
            ftk::KeyShortcut(ftk::Key::Q, ftk::commandKeyModifier),
            [appWeak]
            {
                if (auto app = appWeak.lock())
                {
                    app->exit();
                }
            });

        _documentsObserver = ftk::ListObserver<std::shared_ptr<models::Document> >::create(
            app->getDocumentModel()->observe(),
            [this, appWeak](const std::vector<std::shared_ptr<models::Document> >& value)
            {
                _actions["Close"]->setEnabled(!value.empty());
                _actions["CloseAll"]->setEnabled(!value.empty());
            });
    }

    FileActions::~FileActions()
    {}

    std::shared_ptr<FileActions> FileActions::create(
        const std::shared_ptr<ftk::Context>& context,
        const std::shared_ptr<App>& app,
        const std::shared_ptr<MainWindow>& mainWindow)
    {
        auto out = std::shared_ptr<FileActions>(new FileActions);
        out->_init(context, app, mainWindow);
        return out;
    }

    const std::map<std::string, std::shared_ptr<ftk::Action> >& FileActions::getActions() const
    {
        return _actions;
    }
}